Runtime/Components/CinemachineSplineDolly.cs

```diff
@@ -0,0 +1,371 @@
+            var normalizedSplinePosition = 
+                SplineUtility.ConvertIndexUnit(spline, m_SplinePosition, m_PositionUnits, PathIndexUnit.Normalized);
+            
+            // Init previous frame state info
+            if (deltaTime < 0 || !VirtualCamera.PreviousStateIsValid)
+            {
+                m_PreviousNormalizedSplinePosition = normalizedSplinePosition;
+                m_PreviousCameraPosition = curState.RawPosition;
+                m_PreviousOrientation = curState.RawOrientation;
+            }
+
+            // Get the new ideal spline base position
+            if (m_AutoDolly.m_Enabled && FollowTarget != null)
+            {
+                // convert follow target into spline local space, because SplineUtility works in spline local space
+                SplineUtility.GetNearestPoint(spline, 
+                    m_Spline.transform.InverseTransformPoint(FollowTargetPosition), out _, out normalizedSplinePosition, 
+                    m_AutoDolly.m_SearchResolution, m_AutoDolly.m_SearchIteration);
+                m_SplinePosition = SplineUtility.ConvertIndexUnit(spline, normalizedSplinePosition, 
+                    PathIndexUnit.Normalized, m_PositionUnits);
+                
+                // Apply the spline position offset
+                normalizedSplinePosition += m_AutoDolly.m_PositionOffset;
+            }
+
+            if (deltaTime >= 0 && VirtualCamera.PreviousStateIsValid)
+            {
+                float maxUnit = 1; // we are always using normalized unit [0-1]
+                {
+                    float prev = m_PreviousNormalizedSplinePosition;
+                    float next = normalizedSplinePosition;
+                    if (spline.Closed && Mathf.Abs(next - prev) > maxUnit / 2)
+                    {
+                        if (next > prev)
+                            prev += maxUnit;
+                        else
+                            prev -= maxUnit;
+                    }
+                    m_PreviousNormalizedSplinePosition = prev;
+                    normalizedSplinePosition = next;
+                }
+
+                // Apply damping in the spline direction
+                float offset = m_PreviousNormalizedSplinePosition - normalizedSplinePosition;
+                offset = Damper.Damp(offset, m_Damping.z, deltaTime);
+                normalizedSplinePosition = m_PreviousNormalizedSplinePosition - offset;
+            }
+            m_PreviousNormalizedSplinePosition = normalizedSplinePosition;
+            m_Spline.Evaluate(normalizedSplinePosition, 
+                out var localPosition, out var localTangent, out var localUp);
+            Vector3 newCameraPos = localPosition;
+            if (Vector3.SqrMagnitude(localTangent) != 0)
+            {
+                m_CorrectedTangent = localTangent;
+            }
+            if (Vector3.SqrMagnitude(localUp) != 0)
```

Comment:
is it ok to do hard equalities in these kinds of tests?  Is `m_Spline.Evaluate` expected to output unit vectors or zero?  If so, then document it and do a hard check for zero instead of computing magnitudes for nothing.

Also: are these expected to be either both zero or both nonzero?  If it can be that only one of them is zero then it's possible for them to become parallel in some cases, which is a problem.  This needs to be clarified.

Reply:
You are right, that's better to check for 0,0,0.

Tangent is not unit vector, but it return 0,0,0, when tangent is set to be 0,0,0. Otherwise it returns a correct tangent vector.
UpVector is unit, but it cannot be 0,0,0.